xds/internal/server/listener_wrapper_test.go

```diff
@@ -151,5 +157,109 @@ func (s) TestListenerWrapper(t *testing.T) {
 			t.Fatalf("mode change received: %v, want: %v", mode, connectivity.ServingModeNotServing)
 		}
 	}
+}
+
+type testService struct {
+	testgrpc.TestServiceServer
+}
+
+func (*testService) EmptyCall(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+	return &testpb.Empty{}, nil
+}
+
+// TestConnsCleanup tests that the listener wrapper clears it's connection
+// references when connections close. It sets up a listener wrapper and gRPC
+// Server, and connects to the server 100 times and makes an RPC each time, and
+// then closes the connection. After these 100 connections Close, the listener
+// wrapper should have no more references to any connections.
+func (s) TestConnsCleanup(t *testing.T) {
+	mgmtServer, nodeID, _, _, xdsC := xdsSetupForTests(t)
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create a local TCP listener: %v", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	modeCh := make(chan connectivity.ServingMode, 1)
+	vm := verifyMode{
+		modeCh: modeCh,
+	}
+
+	host, port := hostPortFromListener(t, lis)
+	lisResourceName := fmt.Sprintf(e2e.ServerListenerResourceNameTemplate, net.JoinHostPort(host, strconv.Itoa(int(port))))
+	params := ListenerWrapperParams{
+		Listener:             lis,
+		ListenerResourceName: lisResourceName,
+		XDSClient:            xdsC,
+		ModeCallback:         vm.verifyModeCallback,
+	}
+	lw := NewListenerWrapper(params)
+	if lw == nil {
+		t.Fatalf("NewListenerWrapper(%+v) returned nil", params)
+	}
+	defer lw.Close()
+
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultServerListener(host, port, e2e.SecurityLevelNone, route1)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for Listener Mode to go serving.
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout waiting for mode change")
+	case mode := <-modeCh:
+		if mode != connectivity.ServingModeServing {
+			t.Fatalf("mode change received: %v, want: %v", mode, connectivity.ServingModeServing)
+		}
+	}
+
+	server := grpc.NewServer(grpc.Creds(insecure.NewCredentials()))
+	testgrpc.RegisterTestServiceServer(server, &testService{})
+	wg := sync.WaitGroup{}
+	wg.Add(1)
```

Comment:
Is this necessary?

Does `Serve` block on anything after `Stop` (not `GracefulStop`) is called?

Reply:
No, but I forget the testing requirement for the goroutines not leaking. The signal of Stop causes Serve to eventually exit, but then it's not guaranteed it'll exit before the goroutine returns. Although there really isn't a distinction because you can execute wg.Done() yield testing goroutine returns and goroutine hasn't finished, but I think you mentioned to me a while back that this is ok with respect to leak checker.

Reply:
Deleted this wait group.

Reply:
Actually, all the functionality you're doing is already implemented in the channel itself for its own reasons:

`Serve` does the `wg.Done` when it exits:
https://github.com/grpc/grpc-go/blob/6f50403edb0c2db1bb557168cf4e6f87ea2efdb5/server.go#L851

And `Stop` does the `wg.Wait` (way) before it returns:
https://github.com/grpc/grpc-go/blob/6f50403edb0c2db1bb557168cf4e6f87ea2efdb5/server.go#L1911

(`Stop` needs `Serve` to end in order to be sure it can close _all_ the connections the channel created.)

Reply:
Ah, ok, that's interesting. Noted for next time. I already deleted the wait group and the wait, so I'll go ahead and merge this.